[ONNX] Restore readable names for parameters and buffers

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101029
* #104741
* __->__ #104493


This PR introduces a new pass that restore parameter and buffer names from original module.

It is useful for readability of the exported ONNX graph. It restores the parameter and buffer names from the original module. For example, if the original module has a parameter named `root.linear.0.weight`, and the parameter is renamed to
  `_param_constant9` by FX, this pass will rename it back.

BowenBao committed:
[ONNX] Restore readable names for parameters and buffers

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Restore readable names for parameters and buffers"




[ghstack-poisoned]

wschin commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,111 @@
+
+    This pass is useful for readability of the exported ONNX graph. It restores the
+    parameter and buffer names from the original module. For example, if the original
+    module has a parameter named `root.linear.0.weight`, and the parameter is renamed to
```
Good document always comes with examples! Nice!

wschin commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,111 @@
+                )
+                if isinstance(attr_value, torch.nn.Parameter):
+                    # If it is a parameter we treat it more seriously.
+                    diagnostic.level = diagnostics.levels.WARNING
```
Alter state is not recommended. Maybe keep it untouched?

BowenBao replied:
We are logging content into the diagnostic object throughout the execution of the function. The intent here is to increase the severity of a diagnostic when certain issues were found. Do you have suggestions for a better api?

wschin replied:
`diagnistic.with_local_level(WARNING)`

wschin commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,111 @@
+                    continue
+                attr_value = getattr(self.module, node.target)
+                if (
+                    isinstance(attr_value, (torch.nn.Parameter, torch.Tensor))
```
Why `torch.nn.Parameter`? I thought it's a `torch.Tensor`. No?

BowenBao replied:
`torch.nn.Parameter` is the original form for parameters. This is what differentiates them from buffers.

wschin commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,111 @@
+                    # We have already processed this parameter/buffer.
+                    old_name_to_nodes[node.target][0].append(node)
+                    continue
+                attr_value = getattr(self.module, node.target)
```
Does it work with `node.target="self.h0.h1.h2.weight"`?

BowenBao replied:
Good question. No it does not. However, the example you give should never happen in our current code path. all parameters and buffers are top level after running aten decomp.

I'm putting a check in this PR. I'm not in favor of actually supporting this because 1. this means something unexpected happened at fx generation stage, 2. it adds a lot of unnecessary complexity. I'm open to discussion if you feel strongly otherwise.

thiagocrepaldi replied:
At least document in the pass docstring that a decomposition is assumed to be done before this pass. This creates a order dependency between the pass, but hopefully decomposition will be done by torch._dynamo.expor4t in the future and this dependency will be gone

BowenBao replied:
Ok, added

    This pass must be run after `Decompose` pass. Because this pass is expected to be called on
    `fx.GraphModule` produced by `proxy_tensor.make_fx`, where all parameters and buffers
    are registered at root level.

wschin commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,111 @@
+        )
+        diagnostic = self.diagnostic_context.inflight_diagnostic()
+
+        old_name_to_nodes: Dict[str, Tuple[List[torch.fx.Node], str]] = {}
```
Always explain collection objects with examples. E.g.,
```
# something[i] is the name of the i-th node in graph.
something = [...]
```

BowenBao replied:
done

wschin commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,111 @@
+        state_to_readable_name.update(
+            {v: k for k, v in self.original_module.named_buffers()}
+        )
+        diagnostic = self.diagnostic_context.inflight_diagnostic()
```
What is `inflight_diagnostic`?

BowenBao replied:
`@diagnose_call` creates a diagnostic instance to track the decorated method's arguments, return values and any exceptions raised within. It gets logged into diagnostic context when the function ends, hence the name 'inflight'. `inflight_diagnostic()` retrieves that diagnostic instance.

I will follow up with more diagnostics docs.

wschin commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,111 @@
+        assert (
+            len(kwargs) == 0
+        ), "RestoreParameterAndBufferNames does not take any kwargs"
+        state_to_readable_name: Dict[Union[torch.nn.Parameter, torch.Tensor], str] = {}
```
Please explain every collection object with examples. E.g.,
```
# something[i] is the name of the i-th node in graph.
something = [...]
```

BowenBao commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,111 @@
+        ), "`nodes` must all have same `target`"
+        old_name = nodes[0].target
+        assert isinstance(old_name, str), f"Expected str, got type({old_name})"
+        # Parameter/buffer name cannot contain "."
```
Sure.

BowenBao replied:
Actually I cannot do this in this PR. A bug is revealed that we are naming ONNX initializer by node.name instead of node.target. It has two issues

1. Duplicated initializers if multiple `getattr` nodes were called to retrieve same parameter.
2. `node.name` is not always the same as `node.target`. Previously this is left unnoticed. Uncovered here due to that '/' is not allowed in `node.name`, but allowed in `node.target`. This means we still get `_` in the onnx model initializer names in the end, even if we set to `/` here.

My plan is to merge this PR with `_` to stay consistent, and follow up another PR to fix initializer name, and revert to `/` there.

thiagocrepaldi replied:
I am actually ok with _ because there will always be a single scope, provided this runs after decomposition. Maybe a double _ instead of a single, if we really want to distinguish from "regular" _ on existing names

wschin replied:
`_` is a common token in variables. Since the goal is to perserve as much as the original model structure information, `/` is recommanded.

thiagocrepaldi reviewed (commented):
Are you adding unit test to check names are actually being preserved?

BowenBao committed:
Update on "[ONNX] Restore readable names for parameters and buffers"



This PR introduces a new pass that restore parameter and buffer names from original module.

It is useful for readability of the exported ONNX graph. It restores the parameter and buffer names from the original module. For example, if the original module has a parameter named `root.linear.0.weight`, and the parameter is renamed to
  `_param_constant9` by FX, this pass will rename it back.

[ghstack-poisoned]

BowenBao commented:
@wschin @thiagocrepaldi  updated PR, PTAL.

thiagocrepaldi commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -164,6 +164,42 @@ def forward(self, input):
         )
 
+    def test_dynamo_export_retains_readable_parameter_and_buffer_names(self):
+        class MNISTModel(nn.Module):
```
How about adding a buffer? The PR handles both parameters and buffers, right?

BowenBao replied:
done

wschin commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -164,6 +164,42 @@ def forward(self, input):
+        class MNISTModel(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.conv1 = nn.Conv2d(1, 32, 3, 1, bias=False)
```
Let's test nested custom nn.Module.

wschin reviewed (approved):
LGTM.

BowenBao committed:
Update on "[ONNX] Restore readable names for parameters and buffers"



This PR introduces a new pass that restore parameter and buffer names from original module.

It is useful for readability of the exported ONNX graph. It restores the parameter and buffer names from the original module. For example, if the original module has a parameter named `root.linear.0.weight`, and the parameter is renamed to
  `_param_constant9` by FX, this pass will rename it back.

[ghstack-poisoned]

thiagocrepaldi reviewed (approved):
nice

BowenBao commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,127 @@
+        # the parameter/buffer. E.g., "self.linear.weight".
+        state_to_readable_name: Dict[Union[torch.nn.Parameter, torch.Tensor], str] = {}
+        state_to_readable_name.update(
+            {v: k for k, v in self.original_module.named_parameters()}
+        )
+        state_to_readable_name.update(
+            {v: k for k, v in self.original_module.named_buffers()}
+        )
```
Not sure about this one.
* This is not doing name mangling, but rather constructing a reverse lookup table.
* In previous code the replacement happens near the source of that requirement. Putting it here feels arbitrary. 
* If let's say we do need to support params that already exist with format of `xxx.yyy`, we cannot use this lookup table because "." is already replaced.
* Now `.replace(".", "_")` appears twice, then we need another API / lambda just for that. Feels over designing.

thiagocrepaldi replied:
My point was that `state_to_readable_name` does not store the *final* readable name as it is. It is just a temporary state that will be post processed by yet another function. if we need to know the updated name in another place, we will have to repeat the same search/replace at `state_to_readable_name` to get the final name again

not a blocking though

BowenBao committed:
address comments on "[ONNX] Restore readable names for parameters and buffers"



This PR introduces a new pass that restore parameter and buffer names from original module.

It is useful for readability of the exported ONNX graph. It restores the parameter and buffer names from the original module. For example, if the original module has a parameter named `root.linear.0.weight`, and the parameter is renamed to
  `_param_constant9` by FX, this pass will rename it back.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,127 @@
+        assert len(args) == 0, "RestoreParameterAndBufferNames does not take any args"
+        assert (
+            len(kwargs) == 0
+        ), "RestoreParameterAndBufferNames does not take any kwargs"
```
couldn't do this, lintrunner complains, reverting.

BowenBao committed:
Update on "[ONNX] Restore readable names for parameters and buffers"



This PR introduces a new pass that restore parameter and buffer names from original module.

It is useful for readability of the exported ONNX graph. It restores the parameter and buffer names from the original module. For example, if the original module has a parameter named `root.linear.0.weight`, and the parameter is renamed to
  `_param_constant9` by FX, this pass will rename it back.

[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,127 @@
+
+        For each `get_attr` node, if the target is a str representing a parameter or buffer
+        under `self.module`, we rename the parameter or buffer to its original name.
+        The parameters and buffers between `self.module` and `self.original_module` refer
```
```suggestion
        The parameters and buffers between `self.fx_module` and `self.nn_module` refer
```

BowenBao replied:
I'll fix these in a follow up to merge this first to unblock you.

BowenBao replied:
nvm I think better to rebase to be safe.. Another PR with pass was merged.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,127 @@
+        """Restore parameter and buffer names from original module.
+
+        For each `get_attr` node, if the target is a str representing a parameter or buffer
+        under `self.module`, we rename the parameter or buffer to its original name.
```
```suggestion
        under `self.fx_module`, we rename the parameter or buffer to its original name.
```
there is a bunch of old `self.module` going on here

BowenBao replied:
can't do anything, this is in base class

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,127 @@
+                        f"All parameters and buffers are expected to be registered at root level, "
+                        f"i.e., self.module. "
+                    )
+                if node.target in old_name_to_nodes:
```
just double checking. Does this update case actually happen? 

by reading https://github.com/pytorch/pytorch/issues/104670 bullet 1), it seemed that only `node.name` can be duplicated, but not `node.target` and therefore we would start using `node.target`

BowenBao replied:
haha it's the opposite, `node.name` cannot be duplicated, `node.target` can. 

So one torch parameter can be the `node.target` of many `get_attr` nodes, hence many `node.name`s. Creating initializer based on `node.name`s will give us a bunch of initializers with different names, but they all duplicate the same weight data.

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,127 @@
+        nn_module: torch.nn.Module,
+    ):
+        super().__init__(diagnostic_context, fx_module)
+        self.original_module = nn_module
```
```suggestion
        self.nn_module = nn_module
```

thiagocrepaldi commented on torch/onnx/_internal/fx/passes/readability.py:
```diff
@@ -0,0 +1,127 @@
+        # TODO(bowbao): fix #104670 and replace "." with "/" to avoid collision.
+        normalized_name = new_name.replace(".", "_")
+        attr_value = getattr(self.module, old_name)
+        setattr(self.module, normalized_name, attr_value)
```
```suggestion
        setattr(self.fx_module, normalized_name, attr_value)
```

thiagocrepaldi reviewed (commented):
Besides `fx_module` and `nn_module`, the implementation also created different references for them. `original_module` as another name for `nn_module` and `module` as another name for `fx_name`. The first two names are clearer

BowenBao committed:
Update on "[ONNX] Restore readable names for parameters and buffers"



This PR introduces a new pass that restore parameter and buffer names from original module.

It is useful for readability of the exported ONNX graph. It restores the parameter and buffer names from the original module. For example, if the original module has a parameter named `root.linear.0.weight`, and the parameter is renamed to
  `_param_constant9` by FX, this pass will rename it back.

[ghstack-poisoned]
